### PR TITLE
[Core] Fix A10 GPU on Azure

### LINKED ISSUE #3651: No CUDA drivers in Azure A10
### Bug
- Follow this [guide](https://skypilot.readthedocs.io/en/latest/gallery/llms/qwen.html)
- Expect the vLLM server to come up, but the job fails instead due to no CUDA drivers
- I looked through the Skypilot code but the parts for picking an image get a bit hard to parse. From what I can tell, Skypilot (correctly) picks the `ubuntu-hpc` 22.04 image for a gen 2 instance.

Screenshot 2024-06-07 at 5 03 23 PM
Screenshot 2024-06-07 at 5 03 18 PM

### To Reproduce
1. Run `sky launch -c qwen skypilot.yaml  --cloud azure --region westus3`
2. Observe that the launch fails and errors related to no CUDA drivers found
3. Confirm that CUDA drivers indeed does not exist by `ssh qwen && nvidia-smi`

**skypilot.yaml** (modifed from [qwen-7b.yaml](https://github.com/skypilot-org/skypilot/blob/master/llm/qwen/serve-7b.yaml), with extra logging statements and to use A10 only)
```yaml
envs:
  MODEL_NAME: Qwen/Qwen1.5-7B-Chat

service:
  # Specifying the path to the endpoint to check the readiness of the replicas.
  readiness_probe:
    path: /v1/chat/completions
    post_data:
      model: $MODEL_NAME
      messages:
        - role: user
          content: Hello! What is your name?
      max_tokens: 1
    initial_delay_seconds: 1200
  # How many replicas to manage.
  replicas: 1

resources:
  # accelerators: { L4, A10g, A10, L40, A40, A100:1, A100-80GB:1 }
  accelerators: { A10 }
  disk_tier: best
  ports: 8000

setup: |
  echo "[skypilot.yaml] Activating conda environment 'qwen'"
  conda activate qwen
  if [ $? -ne 0 ]; then
    echo "[skypilot.yaml] Creating new conda environment 'qwen' with Python 3.10"
    conda create -n qwen python=3.10 -y
    conda activate qwen
  fi
  echo "[skypilot.yaml] Installing required packages..."
  pip install -U vllm==0.3.2
  pip install -U transformers==4.38.0
  echo "[skypilot.yaml] Done installing packages."

run: |
  echo "[skypilot.yaml] Listing available conda environments:"
  conda env list
  echo "[skypilot.yaml] Activating conda environment 'qwen'"
  conda activate qwen
  echo "[skypilot.yaml] Listing available conda environments:"
  conda env list
  echo "[skypilot.yaml] Listing installed packages:"
  pip list
  echo "[skypilot.yaml] Setting PATH to include /sbin"
  export PATH=$PATH:/sbin
  echo "[skypilot.yaml] Starting vllm OpenAI API server with the following configuration:"
  echo "[skypilot.yaml]   - Host: 0.0.0.0"
  echo "[skypilot.yaml]   - Model: $MODEL_NAME"
  echo "[skypilot.yaml]   - Tensor Parallel Size: $SKYPILOT_NUM_GPUS_PER_NODE"
  echo "[skypilot.yaml]   - Maximum Model Length: 1024"
  python -m vllm.entrypoints.openai.api_server \
    --host 0.0.0.0 \
    --model $MODEL_NAME \
    --tensor-parallel-size $SKYPILOT_NUM_GPUS_PER_NODE \
    --max-model-len 1024 | tee ~/openai_api_server.log

```

_Version & Commit info:_
* `sky -v`: `skypilot, version 1.0.0.dev20240607`
* `sky -c`: `skypilot, commit 26d902d7e47900bb6b6c897f6fda79047b35df35`

### PR DESCRIPTION
**Pull request by cblmemo**
Fixes #3651. [A10 GPUs require a special type of driver, Grid Driver](https://forums.developer.nvidia.com/t/ubuntu-22-04-installation-driver-error-nvidia-a10/285195/2), which is [suggested to be installed by enabling the `NvidiaGpuDriverLinux` extension](https://learn.microsoft.com/en-us/azure/virtual-machines/extensions/hpccompute-gpu-linux). This PR added this extension for A10 instances.

Tested (run the relevant ones):

- [x] Code formatting: `bash format.sh`
- [x] Any manual or new tests for this PR (please specify below)
     - [x] The YAML in the issue and `nvidia-smi` works well. 
     - [x] The YAML in the issue, Qwen model e2e.
     - [x] `sky launch --cloud azure -c az-wo-a10` and checked that the drivers is not installed
- [ ] All smoke tests: `pytest tests/test_smoke.py` 
- [ ] Relevant individual smoke tests: `pytest tests/test_smoke.py::test_fill_in_the_name` 
- [ ] Backward compatibility tests: `conda deactivate; bash -i tests/backward_compatibility_tests.sh`

### CODE REVIEW

**Comment by cblmemo** [comment]
To discuss: Should we put the template into a separate file like https://github.com/skypilot-org/skypilot/blob/master/sky/skylet/providers/azure/azure-config-template.json

File: sky/skylet/providers/azure/node_provider.py [hunk 1]
```diff
@@ -322,6 +322,73 @@ def _create_node(self, node_config, tags, count):
...
+        # pylint: disable=import-outside-toplevel
+        from sky.clouds.service_catalog import azure_catalog
+
+        instance_type = node_config["azure_arm_parameters"].get("vmSize", "")
+        accs = azure_catalog.get_accelerators_from_instance_type(instance_type)
+        if accs is None or "A10" not in accs:
+            return
+
+        # Configure driver extension for A10 GPUs
+        create_result = poller.result().as_dict()
```
Comment (Michaelvll) [inline sky/skylet/providers/azure/node_provider.py:341]: We should merge the parameters for A10 to original parameters to avoid two `create_or_update` which can cause significant overhead? Please help profile it.

Comment (cblmemo) [reply sky/skylet/providers/azure/node_provider.py:341]: Good point! I benchmarked and it reduces the provision time from `15m15s` to `14m36s`.

**Comment by cblmemo** [comment]
Since now we got more quotas, I also tested the YAML in the issue and it works in e2e as well 🫡

File: sky/skylet/providers/azure/node_provider.py [hunk 2]
```diff
@@ -303,6 +303,37 @@ def _create_node(self, node_config, tags, count):
         template_params["nsg"] = self.provider_config["nsg"]
         template_params["subnet"] = self.provider_config["subnet"]
 
+        # pylint: disable=import-outside-toplevel
+        from sky.clouds.service_catalog import azure_catalog
+
+        instance_type = node_config["azure_arm_parameters"].get("vmSize", "")
+        accs = azure_catalog.get_accelerators_from_instance_type(instance_type)
+        if accs is not None and "A10" in accs:
```
Comment (Michaelvll) [inline sky/skylet/providers/azure/node_provider.py:311]: Is there a way to avoid using `azure_catalog` in the node_provider as that seems to be an abstraction leakage, but I am ok to keep it as is for now.

**Review by Michaelvll — COMMENTED** [review]
Thanks for the update @cblmemo!

File: sky/skylet/providers/azure/node_provider.py (hunk 2 above)
Comment (Michaelvll) [reply sky/skylet/providers/azure/node_provider.py:311]: Actually, how about in the `azure-ray.yaml` we pass a `need_nvidia_driver_extension: true` within `clouds/azure.py::make_deployable_variables`. There we can check the accelerators, and get rid of the use of catalog here.

**Review by Michaelvll — COMMENTED** [review]
Thanks for the update @cblmemo! Please check the comment above. We can get rid of the use of catalog in the `node_provider` with that. Other parts look good to me.

File: sky/skylet/providers/azure/node_provider.py (hunk 2 above)
Comment (cblmemo) [reply sky/skylet/providers/azure/node_provider.py]: Done! PTAL 🫡

File: sky/clouds/azure.py [hunk 3]
```diff
@@ -315,6 +314,10 @@ def make_deploy_resources_variables(
             'image_version': version,
         }
 
+        # Setup the A10 nvidia driver.
+        need_nvidia_driver_extension = (resources.accelerators is not None and
+                                        'A10' in resources.accelerators)
```
Comment (Michaelvll) [inline sky/clouds/azure.py:319]: `resources.accelerators` may not include the accelerators as we may have removed the `accelerators` field once the instance type is set?
```suggestion
        need_nvidia_driver_extension = (acc_dict is not None and
                                        'A10' in acc_dict)
```

Comment (cblmemo) [reply sky/clouds/azure.py]: Good point! Done.

**Review by Michaelvll — APPROVED** [review]
Tried it out and seems working well. Thanks @cblmemo!

The following takes 13mins on my end
```
sky launch --cloud azure --gpus A10 --down nvidia-smi 
```
